src/trie.js

```diff
@@ -1,102 +1,518 @@
+      builder.add(withParam(node.paramed, ''));
+    }
+  }
+  return builder;
+};
+
+// Export execute
+exports.execute = execute;
+
+/**
+ * Create a topological sorting of rules according to dependencies, defined
+ * as follows:
+ *
+ * r1:   A  -> B
+ * r2:   B  -> C
+ * r1 depends on r2
  *
- * Note that, as applied to roles, the PATTERN for every rule begins with
- * `assume:`, but nothing in this file is specific to that prefix.
+ * r1:   A  -> B
+ * r2:   B* -> C
+ * r1 depends on r2
+ *
+ * r1:   A* -> B<..>
+ * r2:   BX -> C
+ * r1 depends on r2
+ *
+ * For any two rules there is a dependency relationship if a parameterization
+ * of one of the rules grants scopes that matches the other scope.
+ *
+ * This returns an topological ordering of rules, where <..> have been replaced
+ * with PARAM in the scopes. If there is a dependency cycle this method will
+ * throw an error with err.code = 'DependencyCycleError' and an err.cycle
+ * property, this is the only effect of this method that useful outside this
+ * file (and testing).
  */
-exports.generateTrie = (rules) => {
-  // generate the state for rules i .. n-1 starting at character position k.
-  // This assumes that the rules have the same value at positions < k, as
-  // assured by the sort order.  Thus we have three segments to work with,
-  // in order:
-  //         k
-  // i -> ...*   -- (A) terminal * in pattern
-  //      ...    -- (B) end of pattern (where pattern[k-1] != '*')
-  //      ...*a  \
-  //      ...a   -- (C) nonterminal * or other characters in pattern
-  //      ...ab  /
-  // n ->
-  let gen = (i, n, k) => {
-    let state = {};
-
-    let j = i;
-    while (j < n) {
-      // find the current segment, at indices [seg .. j-1] or just [j]
-      let current = rules[j].pattern[k]; // current character
-      if (current === '*' && rules[j].pattern.length === k + 1) {
-        // segment type A, terminal * in pattern; there can be only one, so
-        // no need to scan more rules
-        state.scopes = normalized(rules[j++].scopes);
-      } else if (current === undefined) {
-        // segment type B, end of pattern (with pattern[k-1] != '*'); there can
-        // be only one, so no need to scan more rules
-        state['end'] = {scopes: normalized(rules[j++].scopes)};
-      } else {
-        // segment type C, nonterminal * or other characters
-        let seg = j;
-        do {
-          j++;
-        } while (j < n && rules[j].pattern[k] === current);
-        state[current] = gen(seg, j, k+1);
-      }
+const dependencyOrdering = (rules = []) => {
+  rules = transformRules(rules);
+  // For each rule we must have an efficient way to find other rules that it
+  // depends on. To do this we build a trie (because tries are fast), where
+  // matching a rule gives the scope of its index value in the rules array.
+  // This is a bit of a hack, but so long as numbers don't end with '*' the
+  // ScopeSet will not merge them (which would be bad).
```

Comment:
This is kind of gross, but it works :)

Reply:
yeah, before I used a trie.. the topological sorting time was actually dominating the trie generation time.

Especially, for the big cases... It seems crazy, but a naive algorithm has a lot of comparisons to make.